keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1225,6 +1226,97 @@ Note that given the estimation for duration of processing the requests is
 automatically corrected (both up and down), there is no need to change that
 in the initial version.
 
+
+#### Estimate width of LIST request
+
+Terms:
+- `W`: estimated width of the request
+- `N`: number of object(s) estimated to be processed for serving this `LIST` request
+- `L`: value of the user specified `limit` in the request URI
+- `S`: an integer value that represents the semantics of the `LIST` request, we use as a multiplying 
+   factor when we estimate the width.
+
+
+What's not in scope for this iteration:
+- `namespaced` or `cluster scoped`: we will not take this into account while estimating
+  the width of a `LIST`
+- We are not going to use to the Priority Level configuration assigned to this request for width estimation.   
+- Use the watch cache interface to know whether this `LIST` will be served from the cache.
+- The code site (inside p&f server filter) where we inspect the request query parameters for `LIST`
+  is well before these parameters are validated. We will not validate these parameters when we estimate
+  the width. We will ignore the fact that the user might have provided invalid data semantically.
+- We are going to treat `resourceVersion` as opaque, its value will not be used to estimate `N`.   
+
+While determining the width, we keep in sight the following goals:
+- We will inspect the request query parameters and estimate width for the `LIST` request.
+- Be pessimistic, assume the worst. We will possibly overestimate the `width` for certain LIST request.
+
+
+Estimate `S`:
+
+These are the semantics (from https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions)
+- `Most Recent`: served from etcd via quorum read
+- `Not older than`: at least as new as the provided `resourceVersion`
+- `Exact`: at the exact resource version provided
+- `Continue Token, Exact`: return data at the resource version of the initial paginated list call  
+- `Any`: strong consistency is not required, data at any resource version may be served
+
+We want to feed the semantics in to the width estimation process, we can use the following back of the
+envelope mapping:
+```
+S = {'Most Recent'=4 | 'Not older than'=2 | Exact=2 | 'Continue Token, Exact'=2 | 'Any'=1} 
+```
+
+Estimate `N`:
+
+- If the specified value of `limit` is invalid then the apiserver will reject the request? in this case we can
+  set `N=1`.
+- For a valid value of `limit` we know the upper bound of the total number of objects that will be processed, so `N=L`.
```

Comment:
As I mentioned during the last P&F meeting - this isn't really true.

The `limit` is the upper bound for number of `returned objects`, not for the ones to process. In the most pathological case, one can imagine setting limit to 1 and setting a selector to something that doesn't match any object and we will have to process every single object.

So we should use N=L if there is empty selector and #total number of objects otherwise.

Reply:
As far as I understand, traversing the objects and evaluating the selector is not the majority of the CPU cost in a LIST request.  How about we make it `N=f*L`, where `f` is a fraction we determine experimentally?

Reply:
For an individual item it's usually not the majority I agree.

But in a pathological case I described above, if there are e.g. 100,000 pods but my selector is not matching any of them, even if I set limit L=1, then the cost is actually proportional to this 100,000.

So in other words - I would much rather prefer to overestimate (for overload-protection) than underestimate (as this can lead to overloading).